utils/src/arg_parser.rs

```diff
@@ -0,0 +1,582 @@
+
+        self.get_matches(args)?;
+        Ok(())
+    }
+
+    fn validate_matches(&self, args: &[String]) -> Result<()> {
+        for arg_info in self.args_matches.iter() {
+            // The arguments that are marked `required` must be provided by user.
+            if arg_info.required && arg_info.user_value.is_none() {
+                return Err(Error::MissingArgument(arg_info.name.to_string()));
+            }
+            // For the arguments that require a specific argument to be also present in the list
+            // of arguments provided by user, search for that argument.
+            if arg_info.user_value.is_some() {
+                if let Some(arg_name) = arg_info.requires {
+                    if !args.contains(&(format!("--{}", arg_name))) {
+                        return Err(Error::MissingArgument(arg_name.to_string()));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn is_arg_valid(&self, arg: &str) -> Result<()> {
+        if !arg.starts_with(DOUBLE_DASH) {
+            return Err(Error::UnexpectedArgument(arg.to_string()));
+        }
+        let arg_name = &arg[DOUBLE_DASH.len()..];
+
+        // Check if the argument is an expected one.
+        if !self.args_matches.contains(arg_name) {
+            return Err(Error::UnexpectedArgument(arg_name.to_string()));
+        }
+
+        // Check if the argument was not provided more than once.
+        if self
+            .args_matches
+            .get(arg_name)
+            .unwrap()
+            .user_value
+            .is_some()
+        {
+            return Err(Error::DuplicatedArgument(arg_name.to_string()));
+        }
+        Ok(())
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(&mut self, args: &[String]) -> Result<()> {
+        if args.contains(&HELP_ARG.to_string()) {
+            self.print_help();
+            process::exit(0)
+        }
+
+        let mut index = 0;
+        while index < args.len() {
+            let arg = &args[index];
+
+            self.is_arg_valid(arg)?;
+
+            let arg_info = self
+                .args_matches
+                .get(&arg[DOUBLE_DASH.len()..])
+                .ok_or_else(|| Error::UnexpectedArgument(arg[DOUBLE_DASH.len()..].to_string()))?;
+            let arg_value = if arg_info.takes_value {
+                // Check if user provided a value for the arguments that take a value.
+                if index + 1 >= args.len() || args[index + 1].starts_with(DOUBLE_DASH) {
+                    return Err(Error::MissingValue(arg_info.name.to_string()));
+                }
+                &args[index + 1]
+            } else {
+                FLAG_PROVIDED
+            };
+            let updated_arg = arg_info.clone().set_user_value(arg_value.to_string());
+
+            let step = if arg_info.takes_value { 2 } else { 1 };
+            index += step;
+
+            self.args_matches.replace(updated_arg);
+        }
```

Comment:
A more rusty way of doing the same thing, while avoiding the `clone()` operation, would be:

```rust
        let mut iter = args.iter();
        while let Some(arg) = iter.next() {
            self.is_arg_valid(arg)?;

            let arg_info = self
                .args_matches
                .take(&arg[DOUBLE_DASH.len()..])
                .ok_or_else(|| Error::UnexpectedArgument(arg[DOUBLE_DASH.len()..].to_string()))?;

            let val = if arg_info.takes_value {
                iter.next()
                    .filter(|v| !v.starts_with(DOUBLE_DASH))
                    .ok_or_else(|| Error::MissingValue(arg_info.name.to_string()))?
            } else {
                FLAG_PROVIDED
            };

            self.args_matches.insert(arg_info.set_user_value(val.to_string()));
        }
```

Reply:
Looks nice :), changed it.